improver_tests/wxcode/wxcode/test_wxcode_decision_tree.py

```diff
@@ -58,106 +63,118 @@
-        """Test that succeed and fail point to valid values or nodes."""
-        tree = wxcode_decision_tree()
-        for node in tree:
-            succeed = tree[node]['succeed']
-            if isinstance(succeed, str):
-                self.assertEqual(succeed in tree.keys(), True)
-            fail = tree[node]['fail']
-            if isinstance(fail, str):
-                self.assertEqual(fail in tree, True)
-
-    def test_diagnostic_len_match(self):
-        """Test diagnostic fields, thresholds and conditions are same
-        nested-list structure."""
-        tree = wxcode_decision_tree()
-        for node in tree:
-            query = tree[node]
-            check_diagnostic_lists_consistency(query)
-
-    def test_probability_len_match(self):
-        """Test probability_thresholds list is right shape."""
-        tree = wxcode_decision_tree()
-        for _, query in tree.items():
-            check_list = query['probability_thresholds']
-            self.assertTrue(all([isinstance(x, (int, float))
-                                 for x in check_list]))
-            self.assertTrue(len(check_list) == len(query['diagnostic_fields']))
-
-    def test_gamma_len_match(self):
-        """Test diagnostic_gamma list is right shape if present."""
-        tree = wxcode_decision_tree()
-        for _, query in tree.items():
-            check_list = query.get('diagnostic_gamma', None)
-            if not check_list:
-                continue
-            self.assertTrue(all([isinstance(x, (int, float))
-                                 for x in check_list]))
-            self.assertTrue(len(check_list) == len(query['diagnostic_fields']))
-
-
-if __name__ == '__main__':
-    unittest.main()
+@pytest.mark.parametrize('tree_name', TREE_NAMES)
+def test_basic(tree_name):
+    """Test that the wxcode_decision_tree returns a dictionary."""
+    tree = TREES[tree_name]
+    assert isinstance(tree, dict)
+
+
+@pytest.mark.parametrize('tree_name', TREE_NAMES)
+def test_keywords(tree_name):
+    """Test that the only permissible keywords are used."""
+    tree = TREES[tree_name]
+    all_key_words = REQUIRED_KEY_WORDS + OPTIONAL_KEY_WORDS
+    for node in tree:
+        for entry in tree[node]:
+            assert entry in all_key_words
+
+
+@pytest.mark.parametrize('tree_name', TREE_NAMES)
+def test_start_node_in_tree(tree_name):
+    """Test that the start node is in the tree"""
+    tree = TREES[tree_name]
+    start_node = START_NODES[tree_name]
+    assert start_node in tree
+
+
+@pytest.mark.parametrize('tree_name', TREE_NAMES)
+def test_keywords_diagnostic_missing(tree_name):
+    """Test only set keywords are used in diagnostic_missing_action."""
+    tree = TREES[tree_name]
+    all_key_words = KEYWORDS_DIAGNOSTIC_MISSING_ACTION
+    for items in tree.values():
+        if 'diagnostic_missing_action' in items:
+            entry = items['diagnostic_missing_action']
+            assert entry in all_key_words
```

Comment:
This test does not actually apply to the global tree which does not use a diagnostic_missing_action option anywhere. We can either take the parameterisation out or accept that it will run and do nothing.

Reply:
I think I would leave this in place. While there are no diagnostic_missing_action options in the global tree, and this wasn't being tested before this PR, if they are added in the future, it would be difficult to identify that they aren't being tested.

Reply:
After talking to Ben, I've changed my mind. If we add something new like this to the global tree in the future, we should check that it is tested as per any other code change. I have removed the parametrization for this test.